### ISSUE TITLE
Expose sized integer types from Prelude

### ISSUE THREAD

**Issue by chessai** [haskell/core-libraries-committee#156]
I propose to export the following types from the Prelude:
- [Int8](https://hackage.haskell.org/package/base-4.18.0.0/docs/Data-Int.html#t:Int8)
- [Int16](https://hackage.haskell.org/package/base-4.18.0.0/docs/Data-Int.html#t:Int16)
- [Int32](https://hackage.haskell.org/package/base-4.18.0.0/docs/Data-Int.html#t:Int32)
- [Int64](https://hackage.haskell.org/package/base-4.18.0.0/docs/Data-Int.html#t:Int64)
- [Word8](https://hackage.haskell.org/package/base-4.18.0.0/docs/Data-Word.html#t:Word8)
- [Word16](https://hackage.haskell.org/package/base-4.18.0.0/docs/Data-Word.html#t:Word16)
- [Word32](https://hackage.haskell.org/package/base-4.18.0.0/docs/Data-Word.html#t:Word32)
- [Word64](https://hackage.haskell.org/package/base-4.18.0.0/docs/Data-Word.html#t:Word64)

I've been writing Haskell for years and I use these types often. In every other language with types, I don't have to import a special module to use them. But with Haskell, I do, and they're even split across two modules: `Data.Int` and `Data.Word`.

Any fixes should be very simple, because it will usually involving just deleting 1-2 import lines (of `Data.Int` and/or `Data.Word`).

These types are extremely stable and I don't envision them changing anytime soon (basically ever). This makes the chance of breakage after their introduction to the Prelude effectively zero.

**Comment by ocharles** [comment]
> Any fixes should be very simple, because it will usually involving just deleting 1-2 import lines (of Data.Int and/or Data.Word).

While this is true, this is the kind of change that I find exhausting as an industrial user. We have thousands of modules and do use Int64 and friends a reasonable amount. While I'm sympathetic to this change, this will cause a bunch of warnings, and as we compile with `-Werror`, ultimately cause a lot of busy work just to upgrade. I'm not sure it worth it, especially as HLS now auto-adds imports for me

**Comment by chessai** [comment]
> > Any fixes should be very simple, because it will usually involving just deleting 1-2 import lines (of Data.Int and/or Data.Word).
> 
> While this is true, this is the kind of change that I find exhausting as an industrial user. We have thousands of modules and do use Int64 and friends a reasonable amount. While I'm sympathetic to this change, this will cause a bunch of warnings, and as we compile with `-Werror`, ultimately cause a lot of busy work just to upgrade. I'm not sure it worth it, especially as HLS now auto-adds imports for me

HLS has been unreliable for me at work for multiple years, so I only try it every now and then, and then end up not using it, so I've not been able to benefit from these things.

The most common migration path for this change can be implemented with something as simple as `sed`, which is about as simple as it gets. If you aren't using any of the byteSwap functions (which I don't think most users are), the upgrade should be executing:
```
find /path/to/src -type f -name '*.hs' -exec sed -i '/Data[.]Int\|Data[.]Word/d' {} +
```
which might need some fiddling, but the general idea is the same.

As an industrial user, I also find it exhausting to have to import these in multiple projects/files all over the place.

Even if a project uses a custom prelude, many don't, and libraries usually don't (and probably shouldn't most of the time).
A lot of my use is in libraries and not in projects with custom preludes.

**Comment by chessai** [comment]
> > > Any fixes should be very simple, because it will usually involving just deleting 1-2 import lines (of Data.Int and/or Data.Word).
> > 
> > While this is true, this is the kind of change that I find exhausting as an industrial user. We have thousands of modules and do use Int64 and friends a reasonable amount. While I'm sympathetic to this change, this will cause a bunch of warnings, and as we compile with `-Werror`, ultimately cause a lot of busy work just to upgrade. I'm not sure it worth it, especially as HLS now auto-adds imports for me
> 
> HLS has been unreliable for me at work for multiple years, so I only try it every now and then, and then end up not using it, so I've not been able to benefit from these things.
> 
> The most common migration path for this change can be implemented with something as simple as `sed`, which is about as simple as it gets. If you aren't using any of the byteSwap functions (which I don't think most users are), the upgrade should be executing:
> ```
> find /path/to/src -type f -name '*.hs' -exec sed -i '/Data[.]Int\|Data[.]Word/d' {} +
> ```
> which might need some fiddling, but the general idea is the same.
> 
> As an industrial user, I also find it exhausting to have to import these in multiple projects/files all over the place.
> 
> Even if a project uses a custom prelude, many don't, and libraries usually don't (and probably shouldn't most of the time).
> A lot of my use is in libraries and not in projects with custom preludes.

To add another thing, these types are so common in programming that I don't want them to appear in imports, I just expect them to be there. Haskell is the only language which violates that expectation, and I don't see a good reason for it.

As a frequent reader of code, I do traverse the module list with my eyes, and these imports are just noise.

**Comment by ocharles** [comment]
Like I said, I'm sympathetic to the motivations but I just feel like the costs now outweigh the benefits. I look forward to hearing what others think!

**Comment by chessai** [comment]
> Like I said, I'm sympathetic to the motivations but I just feel like the costs now outweigh the benefits. I look forward to hearing what others think!

I do not understand what costs you are talking about; I believe I addressed that the migration will be just running a single shell command (for an entire project) in the vast majority of cases. If there are other costs please mention them.

**Comment by phadej** [comment]
As another anecdotal sample point:

As an industrial user,  I almost always have `TheProject.Base` or `TheProject.Prelude` which often do re-export `Word*` and `Int*` types.

I also often avoid openly re-exporting `Prelude` from project-specific ones, as e.g. sometimes I want `exp` and `log` to be something else than math operations, hide `head` etc.

So this change won't really affect me, neither complicate nor make my life easier at large.

**Comment by Kleidukos** [comment]
Really in favour of this, if only for playground snippets to get less noise in the import statements.

**Comment by ocharles** [comment]
> I'm not sure I understand why this would require fixes at all. Is it to avoid warnings (errors with `-Werror`) about redundant import statements?

Yep

**Comment by hasufell** [comment]
IMO, if you're on `-Werror` you get to keep the pieces.

Any deprecation strategy would break for such users (because it triggers, well, warnings). I don't think we can be expected to follow that standard. 

This is how `-Werror` is expected to work. Otherwise be more fine-grained about your warnings or turn it off.

**Comment by ocharles** [comment]
I'm aware of this and agree with how everything is working, but it's also why I think this change isn't worth it. I don't think warnings are designed to exist forever, people will eventually want to fix the warning, it's just a case of when. I don't really have anything more to add and understand I'm in a minority, but if my position still isn't clear please let me know if I've missed anything.

**Comment by chshersh** [comment]
> Any fixes should be very simple, because it will usually involving just deleting 1-2 import lines (of Data.Int and/or Data.Word).
>
> this will cause a bunch of warnings, and as we compile with -Werror

Any fix will also require `-XCPP` if people support multiple GHC versions **and** compile with `-Werror`. I'm not sure how many such developers and projects there're who do this because it's generally a nightmare in Haskell to support multiple GHC versions warning-free.

But still, this is a lot of work.

<hr>

I want to express similar feelings that this is too much work despite generally being a good change. The Haskell community is mostly volunteers and any approach that requires volunteers to do even more work won't work. One can argue that it'll save time in the future but I'd say there might not be any future if we constantly discourage existing users from using the language.

For me personally, there's so much stuff I want to reexport from `Prelude` that just those `Int*` and `Word*` are not enough while reexporting them creates a lot of work for the entire ecosystem. Burning by churning is real in the Haskell community.

I don't see a good path forward without patching GHC first (e.g. putting warnings for all reexports from `Prelude` under a separate warning flag like `-Wunused-imports-prelude`).

**Comment by phadej** [comment]
> For me personally, there's so much stuff I want to reexport

Indeed, e.g. `Natural`, `Void` (with `absurd` and `vacuous`?), `Coercible` and `coerce` (except Safe Haskell :S), `Type` (from `Data.Kind`).

I don't really know where I would draw the line *for everyone*.

**Comment by davean** [comment]
I use these types all the time but I really can't care if I have to import
them or not. Ideally not but it just doesn't matter.

Has anyone considered marking something as recently re-exported and given
an option to disable duplicate import warnings for things marked for that
period?

**Comment by googleson78** [comment]
The discussion around costs here again reminds me of https://gitlab.haskell.org/ghc/ghc/-/issues/21879.

If the unused imports warning didn't trigger on "duplicate imports" as well, we would possibly not be having this discussion, and not having to "pay the cost" for compatibility, for what is otherwise an all around agreed upon improvement.

**Comment by tomjaguarpaw** [comment]
I'm personally not inclined to make things easier for people who use `-Werror`. If you do that you've made your own bed and now you must lie in it. Warnings are the single most gentle way that we have to communicate API changes to users. I'm inclined to use them for that purpose.

On the other hand I don't see the big deal about exporting these things from the `Prelude`. It doesn't seem like a big deal to import them from somewhere. I've hardly ever used them though (what is everyone else doing that requires them so much?!) so perhaps my radar is misaligned regarding the utility of that.

**Comment by chessai** [comment]
I believe that there are two uses of -Werror to consider:

1. -Werror on your own project
2. -Werror on your own project AND on your deps

If you compile you do (2), you are just asking for pain, and there is not much that can be done to avoid painful dependency upgrades.

If you do (1), this specific kind of change won't force an upgrade of dependencies, because the redundant import won't halt compilation.

**Comment by chessai** [comment]
> Any fix will also require `-XCPP` if people support multiple GHC versions **and** compile with `-Werror`. I'm not sure how many such developers and projects there're who do this because it's generally a nightmare in Haskell to support multiple GHC versions warning-free.

This is true for library code, but how many libraries with wide support windows are actually compiling with -Werror? I'm sure some kowainik ones are, but most other library maintainers have decided that it's just not worth it. So they will require no change.

**Comment by phadej** [comment]
> This is true for library code, but how many libraries with wide support windows are actually compiling with -Werror? 

Trying to keep latest/recent GHC build warnings free is not uncommon practice.

**Comment by Bodigrim** [comment]
For the reference, the existence of `{Int,Word}{8,16,32,64}` in any compliant compiler is mandated by Haskell Report:
* https://www.haskell.org/onlinereport/haskell2010/haskellch18.html#x26-22300018
* https://www.haskell.org/onlinereport/haskell2010/haskellch23.html#x31-26000023

The Report also says that 

>  The Prelude defines only the most basic numeric types: fixed sized integers (Int), arbitrary precision integers (Integer), single precision floating (Float), and double precision floating (Double). Other numeric types such as rationals and complex numbers are defined in libraries. In particular, the type Rational is a ratio of two Integer values, as defined in the Ratio library.

But we already diverged from this by exporting `Word` and `Rational`.

**Comment by hasufell** [comment]
> I've hardly ever used them though (what is everyone else doing that requires them so much?!) so perhaps my radar is misaligned regarding the utility of that.

The argument that convinces me is that it's common practice for stdlibs to expose them.

We're not high level enough that we don't require those types at all.

I know that e.g. streamly uses these types heavily and makes a lot of effort to work efficiently on different platforms.

**Comment by tomjaguarpaw** [comment]
> The argument that convinces me is that it's common practice for stdlibs to expose them.

As a data point, the other language that I use (Python) doesn't seem to (https://stackoverflow.com/questions/31145143/fixed-width-integer-types-e-g-uint32-in-python).

**Comment by ulysses4ever** [comment]
In order for this argument (about other languages) to apply broadly, you'd have to look at languages appeared or significantly revised, say, after 2k. E.g. Rust, Swift, and (my dear) Julia — all have sized integer types in the global scope by default. Big deal was C99: many people try to follow its example (e.g. Swift docs explicitly cite C99 as the source of inspiration).

**Comment by konsumlamm** [comment]
> > The argument that convinces me is that it's common practice for stdlibs to expose them.
> 
> As a data point, the other language that I use (Python) doesn't seem to (https://stackoverflow.com/questions/31145143/fixed-width-integer-types-e-g-uint32-in-python).

Yeah, but Python doesn't have sized integer types in its stdlib in the first place (except for [ctypes](https://docs.python.org/3/library/ctypes.html), but that's for FFI). Its only integer type is `int`, which is the equivalent of `Integer`. The only other languages I can think of that don't have sized integer types in their stdlib are "scripting languages", i.e. Ruby, Perl, Lua, etc. (at least it seems to me that these don't have sized integer types, from a quick search, correct me if I'm wrong), so I don't think that's a useful comparison.

As @ulysses4ever says, basically every modern (statically typed) language has sized integer types in its prelude: Rust, Swift, Go, Kotlin, Scala, F#, Nim, D, Crystal, ...

**Comment by dmcclean** [comment]
I'm opposed to this because these types are fairly arbitrary and that arbitrariness is ugly. I understand that the whole computing universe uses the same arbitrary choice and it has a long history and lots of utility for that reason, but I still don't mind the extra import.

Now if we were to have one Word type indexed by a type level natural, that would be a different story because you'd be importing the one principled concept instead of the list of arbitrary but useful specific things. But machinery to make that work efficiently so that Word16 and friends could just be aliases is probably only a dream.

**Comment by recursion-ninja** [comment]
I am fully in support of this proposal and, the past of sounding a bit radical, completely on sympathetic to the maintenance concerns raise by industrial users.

**Comment by santiweight** [comment]
Such a change to Prelude should be largely trivial. It is a purely structural change with little to no conflicts (`Word8` is not a commonly user-defined type).

In order to push this through, the bare minimum should include: 
 1. a newly-added-to-cabal `cabal migrate ghc-9.4 ghc-9.6`, which transforms your code with the relevant changes.
 2. `cabal migrate ghc-9.4 ghc-9.6 --backwards-compatible` which would at least _attempt_ to include the appropriate backwards-compatible CPP
 3. (optional) a tool to migrate hackage libraries (that are public) automagically so that maintainers just have to approve the change.

Haskell is a language that takes the principled approach. 

**Breaking changes that put the onus on _all_ users is not a community worth standing by.** The right approach is to push for the right language (this proposal is a good idea) whilst not hurting the people meant to benefit from the change.

**Comment by jberryman** [comment]
I agree with @santiweight : a breaking change (even if that's just for users who compile with `Werror`) should come with either a very strong motivation, or an automatic fix (a tool that automatically migrates, and PRs to all of hackage). It would be awesome for the community if we had that kind of tooling and infrastructure.

**Comment by hasufell** [comment]
Introducing a compiler warning is not a breaking change. Not sure where this idea comes from, but it doesn't align with any standard or common practice.

**Comment by athas** [comment]
I think Haskell would have been a (very slightly) better language if these types had always been exported from the Prelude.  The question to me is whether the cost of adding them *now* is greater than the slight benefit.

It seems to me that the main cost is additional warnings, which is mostly a problem for people who use `-Werror`.  I assume people who use `-Werror` only do so in settings where compiler and library versions do not change frequently (that is my situation), so if new warnings are added, fixing them is not *urgent*.  Also, the warnings added by this change are relatively easy to fix.  As a stopgap, one can always just disable the redundant-import warnings.

Some commenters above suggest that a change like this is only acceptable if an automatic porting tool is also made available.  It would be nice to have such a tool, but I find it unlikely that it will ever happen.  Requiring it as a pre-requisite for this proposal is equivalent to rejecting it.  I don't think further discussion along that line is productive, unless someone is independently motivated to work on such tooling (or something exists that I have missed).

Overall I am mildly in favour of this proposal.  Churn is real, but at least this does not *break* code.

**Comment by dpwiz** [comment]
> Any fix will also require -XCPP if people support multiple GHC versions and compile with -Werror.

In retrospect, a big chunk of my maintenance patches are like this. Especially if a package doesn't change too much and attempts to have a deep support well. With time, such packages will collect `#ifdef MIN_PACKAGE_VERSION_base` like mussels, or attach themselves to `acme-everything-compat`, custom preludes or thatnot - all while changing nothing in their core code. This is somewhat sad.

The "no deeper than 3 major GHCs" policy can limit amount of mussels, but the churn is still there.
And I'd be happy to support GHC 3.11 if that would be free (unless the drop in syntactic sugar levels is unbearable).

> The discussion around costs here again reminds me of https://gitlab.haskell.org/ghc/ghc/-/issues/21879.

This indeed looks like a prerequisite for changes like this.

**Comment by dpwiz** [comment]
I think we should rely more on external tooling for linting packages. Or have a separation of "you're loading a footgun here" warning and "compile-time release notes".

**Comment by simonmar** [comment]
> I'm personally not inclined to make things easier for people who use -Werror. If you do that you've made your own bed and now you must lie in it. Warnings are the single most gentle way that we have to communicate API changes to users. I'm inclined to use them for that purpose.

I've arrived at a slightly different position on this. In a large shared codebase that many people work on, warnings don't really work as intended. Without `-Werror`, warnings accrue over time and you end up ignoring all the warning spew during the build because it mostly relates to code that you're not working on. Warnings that appear in code you're not changing will generally not get fixed, because nobody sees the warning: it was part of the output that got discarded when the build happened in the cloud and got cached, and even if you do see the warning, again it's in code that you're not working on so you'd probably ignore it.

So if you want warnings to be useful you're forced to globally enable `-Werror` and keep the whole codebase clean, or else forget about warnings entirely and turn them off. Warnings that just print stuff aren't useful.

In the codebase I work on we use `-Werror` because we want to keep everything clean. But that does mean that any changes in warnings are effectively breaking changes. The poor soul landed with the task of upgrading the compiler gets to fix everything. Yes it's an unforced choice to keep things warning-clean: we don't have to do that. Indeed we treat HLint warnings slightly differently and an HLint failure doesn't break the build. But the main point I want to make here is that for some groups of users, warnings aren't working for gently signalling upcoming breaking changes.

**Comment by hasufell** [comment]
@simonmar I'm afraid I find that to be an insufficient argument and to me this highlights a problem of user workflow that CLC has nothing to do with.

You can:

- return compile time warnings as artifacts in CI that can be downloaded/reviewed
- send automatic emails to the involved teams/code owners
- allow warnings during development and disallow them during releases

Whether you do any of that or not isn't really a concern of CLC or library maintainers.

Warnings are a method of communication. If you ignore it, don't blame upstream.

**Comment by chshersh** [comment]
> I'm afraid I find that to be an insufficient argument and to me this highlights a problem of user workflow that CLC has nothing to do with.

I don't view CLC as a simple machine that says either "LGTM" or "needs impact assessment because of breaking changes". We're humans, and we're supposed to work with other humans to make the language better for everyone else.

It looks hostile and off-putting when people share their real concerns and pain points about this particular change, and we just dismiss them by saying "You've chosen your life, live with it, not our problem". CLC is trusted with taking ownership of the entire problem and finding the best solution to steer the community towards a better situation. We're not supposed to lack as another bureaucratic entity that just sends people to another department by saying "Not our responsibility".

I can echo @simonmar concerns: in any real Haskell project, you eventually accumulate warnings unless you enforce `-Werror`. Saying to those people that we don't care about their DX sends the same vibes as _"Your values are wrong, you're not using the language in a way I use"_ and this shows high lack of empathy.

@hasufell, all your suggestions to deal with warnings won't help keeping the project warning-free unless you have a team of people who highly care about removing all warnings **and** have the capacity (either time and budget for industrial projects or motivation and time for free projects). And this is a really high requirement for any project.

<hr>

Also, I would like to highlight that the real problem here is `-Wunused-imports`. It hides two different behaviours under the same flag. So the real path forward here is to patch `-Wunused-imports` as proposed in [GHC #21879](https://gitlab.haskell.org/ghc/ghc/-/issues/21879).

CLC already had the same problem with `-Wdeprecations` when two completely different warning classes are hidden behind a single flag. We already know how to resolve this problem, we even have a previous case and we don't need a new GHC language pragma for this.

Users of `-Werror` have chosen to enable this flag but they still have the flexibility to enable/disable specific warnings.
We just need to give them a little bit more flexibility.

**Comment by ramirez7** [comment]
Doesn't the argument against this basically amount to "never add anything commonly-imported to Prelude because it breaks Werror users"? Should Prelude really be calcified like that?

**Comment by jberryman** [comment]
@ramirez7 I would say yes, that causes problems for no real gain and should always be avoided, unless there's another motivation than "I use this a lot". Other reasons this is bad:

- people will write tutorials using new ghc that won't compile on older ghc (another pointless thing requiring explanation for newcomers)
- people will write libraries that won't work on old ghc, and will then require CPP around imports

**Comment by NicolasT** [comment]
> Doesn't the argument against this basically amount to "never add anything commonly-imported to Prelude because it breaks Werror users"? Should Prelude really be calcified like that?

IMHO, yes. What's exported from `base`'s `Prelude` is already quite... random, with plenty of things in there which may be barely used, and things not in there which are very frequently used, though all of that depends on the project at hand, as well as the developers who may have certain preferences how to implement some functionality.

I think the CLC should be *very* careful when adding things to `Prelude`, not because of the `-Werror` situation or whatnot, but simply because it's a very slippery slope. Good arguments can be made for *many* things which, according to some, should become part of the default namespace/`Prelude`. This proposal is, IMHO, one of those.

My personal stance: I don't think `IntN` should be in `Prelude`. They're trivial to `import`, aren't used that often in my experience except for very specific code/modules (where they're just an `import` away), would clutter `Prelude` even more than it already is, and other languages having them "by default" doesn't sound like a convincing argument to me (e.g., C doesn't).

**Comment by ramirez7** [comment]
But isn't Prelude just re-exports of `base`? I'm not understanding where this mythical "change to Prelude that is more impactful than adding new default imports" is going to be found?

Is it just that it's now time to freeze `Prelude`'s exports because there's now a critical-enough mass of users who would be bothered by it and therefore -1 a proposal like this one?

Also fwiw, explicitly sized numbers aren't just for super specific modules. I'd say the majority of Haskell apps I write use them, and I do run-of-the-mill software dev + some hobby stuff:
- I'd say most web apps and services will use them since there are sized numbers in their SQL database (I know in Golang I constantly use the sized variants because it encourages good defaults by making them available). 
- Interfacing with gamedev/gfx/other C libraries is another where I constantly use sized variants. 

I guess you could argue the domain libraries should re-export, but isn't the argument against that change the same as the argument here. That it'll cause breakage for consumers of popular libraries, invalidate tutorials, etc.

**Comment by tomjaguarpaw** [comment]
I think the important issue here is not whether one should use `-Werror` at all, but whether redundant import warning should be caught by that. My opinion is that they shouldn't for two reasons:

1. It's not indicative of a severe enough problem (unlike, say, `-Wincomplete-patterns`), and
2. You're putting a barrier in the way of your library providers of migrating to better module structures (case in point, this proposal)

**Comment by ramirez7** [comment]
And thinking about this argument further, it feels a bit circular.

The only `base` modules that would make sense to add to `Prelude` are commonly-imported ones.

But commonly-imported modules have more widespread breakage.

**Comment by Bodigrim** [comment]
As a point of order, let's stick to the topic and not digress into general discussion.

I do have a huge codebase with `-Werror`, but I downgrade `-Wwarn=unused-imports`, and I strongly recommend others to do the same. This is a harmless warning, not pointing to any potential deficiency or bug (as unused variables or unexhaustive patterns do).

If you are that pedantic that downgrading `-Wunused-imports` is not an option for you, I suggest to be consistent and follow PVP pedantically as well. PVP does recommend importing identifiers explicitly or using a qualification, which would shield you from any issues with additive extension of `Prelude`.

Similar to others above, I agree with https://gitlab.haskell.org/ghc/ghc/-/issues/21879: the real issue is that this warning is unhelpful. Instead of bleeding each time someone adds a re-exported identifier, people insisting on `-Werror` should rather direct efforts to change the behaviour of the warning and eliminate this pain forever.

(To be clear: I am concerned about uncontrolled expansion of `Prelude` and far from certain about my vote on the proposal)

**Comment by aaronallen8455** [comment]
The [pinned-warnings](https://hackage.haskell.org/package/pinned-warnings)  plugin can automatically fix redundant import warnings. It could potentially serve as a migration tool for a change such as this.

**Comment by endgame** [comment]
For whatever my 2¢ is worth: my gut reaction to this proposal is "yes", because of the "every other language has these types to hand". But that argument is insufficient on its own, and I have argued against other proposals based on it (e.g. adding operators like `(<|)` and `(|>)`). So then I am left wondering what general principle I'd endorse wrt proposals to change `Prelude`. I would probably find myself arguing against proposals to remove items from `Prelude` to avoid breaking changes. Does that mean `Prelude` gets to accumulate endless barnacles? Custom preludes fix this, but they are inappropriate for libraries. Project-specific preludes fix that, but mean you're copy-pasting the same custom prelude everywhere. Should we then have selectable preludes/prelude editions that ship with the compiler? That brings us full-circle to the old "education prelude vs. industrial prelude" discussions.

I think I've argued myself away from my gut reaction and am now +ε on the idea.

**Comment by friedbrice** [comment]
> I've been writing Haskell for years and I use these types often. In every other language with types, I don't have to import a special module to use them. But with Haskell, I do

I see what you're saying, but what I'm hearing is that we need qualified namespace re-export...

**Comment by chessai** [comment]
> > I've been writing Haskell for years and I use these types often. In every other language with types, I don't have to import a special module to use them. But with Haskell, I do
> 
> I see what you're saying, but what I'm hearing is that we need qualified namespace re-export...

Can you elaborate on this?

**Comment by chessai** [comment]
> I'm opposed to this because these types are fairly arbitrary and that arbitrariness is ugly. I understand that the whole computing universe uses the same arbitrary choice and it has a long history and lots of utility for that reason, but I still don't mind the extra import.
> 
> Now if we were to have one Word type indexed by a type level natural, that would be a different story because you'd be importing the one principled concept instead of the list of arbitrary but useful specific things. But machinery to make that work efficiently so that Word16 and friends could just be aliases is probably only a dream.

This overlooks the fact that this proposal is forwards-compatible with the API that you're proposing. If we one day have this Nat-indexed Word machinery, we could simple export type synonyms `type Word8 = MyIdealWord 8`.

Also, this machinery does not exist in base, and would be more work to design.

**Comment by jberryman** [comment]
@Bodigrim 

> As a point of order, let's stick to the topic and not digress into general discussion.

Well I think CLC needs a plan for approaching changes like this, unless one wants a repeat of the disagreements here. Luckily I think your comment suggests the way[^1]:


- alter `-Wall` (or maybe `-Wunused-imports` itself) so it ignores cases like would come up with this change
- probably: give a bit of guidance about valid reasons for moving things around in `base`, or into Prelude; make it known what data the proposer should provide; are custom preludes the solution?; should all types whose primary interface are Prelude classes be moved into Prelude?; etc (to get specific: this current proposal is weak: no data about how often `Data.Int/Word` is imported, or how this compares to e.g. other modules or functions/types; no survey of hackage to see how many packages would break, perhaps it's a very small number?)
- proceed with this sort of thing (personally I'd be okay with this proposed change in that case, even though I think [there are other costs](https://github.com/haskell/core-libraries-committee/issues/156#issuecomment-1509058859))

It should be the responsibility of CLC to plan these sorts of transitions I think.

> PVP does recommend importing identifiers explicitly

Maybe I'm misunderstanding you, but iiuc any `import Data.Word (Word8)` would trigger `-Wunused-imports` after this change, like:
```
{-# OPTIONS_GHC -Wall #-}
import GHC.Num(Num)
```

[^1]: ...but otherwise reads as dismissive to me; keep in mind, at least for me, when I see a proposal like this I'm not thinking of just code I help maintain, but also libraries I'll have to help update with PRs, some of which may be in "maintenance mode", i.e. perhaps working well but without an active maintainer. In my opinion keeping the latter kinds of packages building as far as possible is very important, particularly for a community as small as ours

**Comment by mixphix** [comment]
I think it's a hilarious oversight that this was elided from the Prelude when the language was designed. These are machine types for writing high-performance programs.

---

> > I've been writing Haskell for years and I use these types often. In every other language with types, I don't have to import a special module to use them. But with Haskell, I do
> 
> I see what you're saying, but what I'm hearing is that we need qualified namespace re-export...

I wrote [a post on the Discourse](https://discourse.haskell.org/t/fixing-the-prelude-problem/5900) a while ago about a potential solution to qualified re-exports by offering the ability to define an import list from the cabal file. But this misses the point.

The real problem is the PVP.

You can argue all you want that this is a breaking change. But there is no semantic difference to the code. The types are the same, they just live in different places. The only thing preventing a quality-of-life upgrade for programming in Haskell in 2023 is the fact that nobody wants to put in the work to update their codebase. This proposal is a two-line fix in each module. It's really not that much work -- but pile it on top of all the other things that get tossed into an update and all of a sudden you've got a whole bunch of errors to fix.

Haskell needs a migration tool. GHC releases need to be smaller and more frequent. The Prelude desperately needs a major revision. The CLC can do nothing about any of these things on its own, and continues to do nothing without community involvement in DOING THE WORK NECESSARY TO MAKE SHIT BETTER.

**Comment by hasufell** [comment]
> @hasufell, all your suggestions to deal with warnings won't help keeping the project warning-free unless you have a team of people who highly care about removing all warnings and have the capacity (either time and budget for industrial projects or motivation and time for free projects). And this is a really high requirement for any project.

I think you're forgetting that most core library maintainers are volunteers and that never introducing warnings is a huge burden for the library maintainers too.

**Comment by Bodigrim** [comment]
@jberryman you are making interesting points, but I do feel them too general for this particular proposal and the specific kind of toll it introduces. 

> * alter `-Wall` (or maybe `-Wunused-imports` itself) so it ignores cases like would come up with this change

Such change does not fall under CLC purview. That's why I urged those who feel strongly about `-Wall -Werror` to raise a GHC proposal and make their lifes better. By default CLC members vote under assumption that the rest of the ecosystem maintains status quo. 

----

There are several areas of improvement which I would like to see to be able to make an informed decision:

* A brief comparison with other languages. Which languages provide sized integer types from `Prelude`? Which languages have sized integer types, but require additional imports?
* How pervasive is the use of sized integer types in Haskell packages? Which segments of the ecosystem are impacted the most? I imagine packages dealing with FFI?..
* An impact assessment using `clc-stackage` is due. It is unlikely, but possible that a package defines its own `Word8` type or constructor, in which case it will be broken (not just a warning, but name clash error).

**Comment by Bodigrim** [comment]
@chessai just a gentle reminder to elaborate the proposal and prepare impact assessment.

**Comment by KovalevDima** [comment]
@chessai
Would it be possible to resume discussion of this issue?

For example, in my [ClickHouse client](https://github.com/KovalevDima/ClickHaskell/blob/fdf8756c5735041f82c130816c2b9f9fc1c89393/library/ClickHaskell.hs#L66-L67) I currently have to re-export fixed-width integer types:
`Data.Int (Int8, Int16, Int32, Int64)`
`Data.Word (Word8, Word16, Word32, Word64)`.

It would be nice to avoid this in the future

**Comment by chessai** [comment]
> @chessai
> Would it be possible to resume discussion of this issue?
> 
> For example, in my [ClickHouse client](https://github.com/KovalevDima/ClickHaskell/blob/fdf8756c5735041f82c130816c2b9f9fc1c89393/library/ClickHaskell.hs#L66-L67) I currently have to re-export fixed-width integer types:
> `Data.Int (Int8, Int16, Int32, Int64)`
> `Data.Word (Word8, Word16, Word32, Word64)`.
> 
> It would be nice to avoid this in the future

I think that this is a beneficial and wholly uncontroversial change. The maintenance burden is also essentially nonexistent now (down from low) with an agentic invocation, trivially refactoring any executable or libary code's imports.

If the CLC wants to vote on it then they can.
